### PR TITLE
Closes #9004 - Fix issue with access tokens claims not being imported using OIDC IDP Attribute Mappers

### LINKED ISSUE #9004: Access Token claims not imported using OpenID Connect v1.0 Identity Provider Attribute Importer Mappers
### Describe the bug

Hi there !

We have configured an external OIDC Identity Provider (type: `OpenID Connect v1.0`), and are trying to map claims present in the IDP Access Token to Keycloak User Attributes, using an Attribute Importer Mapper.

But Keycloak seems unable to find the claims in the Access Token, with the following DEBUG messages reported:

```
DEBUG [org.keycloak.broker.oidc.mappers.AbstractJsonUserAttributeMapper] (default task-6) Going to process JsonNode path role-coh-irn71429 on data null
DEBUG [org.keycloak.broker.oidc.mappers.AbstractJsonUserAttributeMapper] (default task-6) Going to process JsonNode path iat on data null
```

### Version

10.0.1, but also reproducible with 15.0.2

### Expected behavior

The tooltip message on the Mapper type states:

```
Import declared claim if it exists in ID, access token or the claim set 
returned by the user profile endpoint into the specified user property or attribute.
```

Using the documentation provided [here](https://www.keycloak.org/docs/latest/server_admin/#retrieving-external-idp-tokens), I managed to retrieve the external IDP Token, and can confirm it contains the expected claim. 
So I would actually expect Keycloak to resolve the claims present in the Access Token.

### Actual behavior

The user created in Keycloak at first login does not contain the attribute expected.

### How to Reproduce?

- Configure Keycloak by adding an Identity Provider of type `OpenID Connect v1.0` : https://www.keycloak.org/docs/latest/server_admin/#_identity_broker_oidc
- Configure the external IDP to include a claim in the Access Token only
- In Keycloak, add an Attribute Importer mapper in the identity provider configured. This Attribute Importer should map the claim present in the Access Token
- Try to login using this new Identity Broker. Authentication should pass, but in Keycloak, the user attributes will not contain the mapped claim.

By the way, to make things clearer, I added two test cases in the [following branch in my fork](https://github.com/rm3l/keycloak/tree/claims_not_imported_using_oidc_idp_attribute_mappers_test). 

You can run these tests with the command below:
```shell
mvn -f testsuite/integration-arquillian/pom.xml clean install \
  -Dtest=KcOidcAccessTokenOnlyClaimsUserAttributeMapperTest,OidcAccessTokenOnlyClaimsUserAttributeMapperTest
```

`KcOidcAccessTokenOnlyClaimsUserAttributeMapperTest`, which passes, registers a `keycloak-oidc` Identity Provider.
But as reported here, `OidcAccessTokenOnlyClaimsUserAttributeMapperTest` does not pass because it registers an `oidc` IDP.

### Anything else?

Few months ago, I opened a discussion [here](https://github.com/keycloak/keycloak/discussions/8462) about that issue,but got no response so far. After some discussions, I am finally about to contribute a PR for this issue.

While the reproduction steps are a bit different, @galzetta reported a similar issue (#8961) with similar debug logs. I could not add a test case for this other issue, so I am not sure the linked PR will fix that issue.

### PR DESCRIPTION
**Pull request by rm3l**
This closes #9004.
See #9004 for more context and details about the issue.
Thanks.

### CODE REVIEW

**Comment by bernhardriegler** [comment]
Any updates on fixing the failing tests?

**Comment by rm3l** [comment]
> Any updates on fixing the failing tests?

No idea, as I haven't looked at this PR since then..
I've rebased the branch onto the upstream branch. Let's see how it goes now.

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java [hunk 1]
```diff
@@ -252,8 +252,9 @@ private String getIDTokenForLogout(KeycloakSession session, UserSessionModel use
     }
 
     protected void processAccessTokenResponse(BrokeredIdentityContext context, AccessTokenResponse response) {
-
-
+        // Don't verify audience on accessToken as it may not be there. It was verified on IDToken already
+        JsonWebToken access = validateToken(response.getToken(), true);
+        context.getContextData().put(KeycloakOIDCIdentityProvider.VALIDATED_ACCESS_TOKEN, access);
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:257]: @rm3l The access token in OIDC (it derives from OAuth2) is not format specified in contrast to ID Token which should be a JWT. So an OIDC complaint broker can return Access Tokens that are not a JWT but opaque  (see for example [this example in the spec itself](https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse) which shows an opaque access token and a JWT ID token).

Therefore the mapper only applies to the ID token in the `OIDCIdentityProvider` but it applies to both in the `KeycloakOIDCIdentityProvider` (as keycloak also uses a JWT access token). We cannot assume that any OIDC providers are going to issue JWT access tokens as you are doing here.

Maybe adding an option in the OIDC provider to mark the access token is a JWT is the solution (false by default), but first we need to know if this is really going to be used by people. What OIDC broker are you using? Do you know many implementations out there using JWT tokens as access tokens as keycloak?

**Review by rmartinc — CHANGES_REQUESTED** [review]

**Review by rmartinc — CHANGES_REQUESTED** [review]

**Comment by mbuotidem** [comment]
@rm3l Any plans to incorporate the feedback from @rmartinc?

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java (hunk 1 above)
Comment (mbuotidem) [reply services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:257]: @rmartinc Azure AD is one such IDP that sends the access token as a JWT : https://learn.microsoft.com/en-us/azure/active-directory/develop/access-token-claims-reference so this affects any Keycloak users hoping to broker from Azure AD.

**Comment by rm3l** [comment]
> @rm3l Any plans to incorporate the feedback from @rmartinc?

@mbuotidem @rmartinc I am no longer working with Keycloak. It took very long to get the initial review on this, and meanwhile, I've switched to a completely different project. Please feel free to either take this over and make the requested changes yourselves or close the PR if it is no longer relevant.
I think Keycloak maintainers should have write access to this branch, but if anyone else is interested in continuing this PR, please ping me so that I can give the appropriate write permissions to the branch.
Thanks for your understanding.

**Comment by rmartinc** [comment]
Superseded by #21627. I added the config option over @rm3l initial PR.
Closing this one.

**Pull request closed without merging**